dandiapi/api/services/metadata/__init__.py

```diff
@@ -0,0 +1,106 @@
+from celery.utils.log import get_task_logger
+import dandischema.exceptions
+from dandischema.metadata import validate
+from django.db import transaction
+from django.utils import timezone
+import jsonschema.exceptions
+
+from dandiapi.api.models import Asset, Version
+
+logger = get_task_logger(__name__)
+
+
+def _encode_pydantic_error(error) -> dict[str, str]:
+    return {'field': error['loc'][0], 'message': error['msg']}
+
+
+def _encode_jsonschema_error(error: jsonschema.exceptions.ValidationError) -> dict[str, str]:
+    return {'field': '.'.join([str(p) for p in error.path]), 'message': error.message}
+
+
+def _collect_validation_errors(
+    error: dandischema.exceptions.ValidationError,
+) -> list[dict[str, str]]:
+    if type(error) is dandischema.exceptions.PydanticValidationError:
+        encoder = _encode_pydantic_error
+    elif type(error) is dandischema.exceptions.JsonschemaValidationError:
+        encoder = _encode_jsonschema_error
+    else:
+        raise error
+    return [encoder(error) for error in error.errors]
+
+
+def validate_asset_metadata(*, asset_id: int) -> None:
+    logger.info('Validating asset metadata for asset %s', asset_id)
+    asset: Asset = Asset.objects.get(id=asset_id)
+
+    # Published assets are immutable
+    assert not asset.published
```

Comment:
Is a runtime assertion the right way to do this? Why not raise a customized exception or something else that can carry some semantic information?

Reply:
Yes, we should have a customized exception here, this was a temporary solution until we determined whether or not this check makes sense